### PR TITLE
refactor(store): improve store types

### LINKED ISSUE #732: produce causes TS error
**Describe the bug**
Valid usage of `produce` causes TS error. 

**To Reproduce**
Provide a code snippet and steps to reproduce the behavior:
1. Go to [playground](https://playground.solidjs.com/?version=1.2.2#NobwRAdghgtgpmAXGGUCWEwBowBcCeADgsrgM4Ae2YZA9gK4BOAxiWGjIbY7gAQi9GcCABM4jXgF9eAM0a0YvADo1aAGzQiAtACsyAegDucAEYqA3EogcuPfr2ZCouOAGVc3OFl6F5I+qxSsvKKKnQa2nr6ZB5CFlZWMvQQzLhotBC8AMIMEC6MABQAlPxWvA4ZMbzAMZ7eZHC47p4AurwAvA5OLs1CBSDMubiIvAAMkkWWmRUQVRiOcPB5HbzFHQB8pdPlDU2xcAW+tP6sBWQbW+VXvGQAdIPJuADUT2VXE5Nvkm9CuEyZBTe5QAPCZ6LgPJkCMR2iowRCMipeBkshpmABrdogeZCJa4STrIFXEC1IT3Ibfba8YH6eGQwnTT4QSlWISicQFNbtTbAnKPcS8fTrbwiWjMeh424Ac0aAFE1IthLgAEL4ACSIkBYCghEIKiKTLAkhaQA)
2. Wait for code to be type checked 
3. See the error

**Expected behavior**
No error

**Reproduction**
See above

**Additional context**
The error is silenced in the tutorial by `// @ts-nocheck` directive: https://www.solidjs.com/tutorial/stores_mutation?solved

### PR DESCRIPTION
**Pull request by otonashixav**
store.ts
`NotWrappable` now consists of all non-object/array types.
Simplified `DeepReadonly<T>`.
Simplified `StoreNode`.
`Store<T>` is now an alias for `DeepReadonly<T>`.
Refactored `Part`, `Next`, and `SetStoreFunction` to improve their performance.
Changed `NullableNext` to `Next`, and `Next` to `WrappableNext`. `WrappableNext` excludes all non-wrappable types instead of just non-nullable types, so `{a: number | {b: number}}` now allows `setStore("a", "b", 1)`.
Improved the rest overload for `SetStoreFunction<T>`: it now correctly infers all types in the correct order, except for `ArrayFilterFn`s which need to be explicitly typed.

store.spec.ts
Added a few type test cases for store types.

modifiers.ts
Improved types of `reconcile` and `produce`; the latter now correctly infers types.

modifiers.spec.ts
Allow `produce` calls to have their types inferred.

server.ts
Import types from store.ts instead of redefining them.

mutable.ts
Return `T` instead of `Store<T>`, as the latter is no longer writable.

index.ts
Removed the `Readonly` export. Added `WrappableNext`.

fixes #732

### CODE REVIEW

File: packages/solid/store/src/store.ts [hunk 1]
```diff
@@ -235,141 +215,208 @@ export function updatePath(current: StoreNode, path: any[], traversed: (number |
...
-  index: number
-) => boolean;
+export type ArrayFilterFn<T> = (item: T, index: number) => boolean;
 
-export type Part<T> = T extends any[]
-  ? keyof T | Array<keyof T> | ArrayFilterFn<T> | StorePathRange
-  : T extends object
-  ? keyof T | Array<keyof T>
-  : never;
+export type Part<T> = [T] extends [NotWrappable]
```
Comment (bgoscinski) [inline packages/solid/store/src/store.ts:230]: Out of curiosity: do you have any article or could you tell me what should I search for to understand this, emm... thing?
`[T] extends [U]` 🤯

Comment (bgoscinski) [reply packages/solid/store/src/store.ts:230]: Nvm. I think I found it https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#distributive-conditional-types :)

File: packages/solid/store/src/store.ts [hunk 2]
```diff
@@ -235,141 +215,208 @@ export function updatePath(current: StoreNode, path: any[], traversed: (number |
...
-    K6 extends Part<Next<Next<Next<Next<Next<T, K1>, K2>, K3>, K4>, K5>>,
-    K7 extends Part<Next<Next<Next<Next<Next<Next<T, K1>, K2>, K3>, K4>, K5>, K6>>,
-    K8 extends Part<Next<Next<Next<Next<Next<Next<Next<T, K1>, K2>, K3>, K4>, K5>, K6>, K7>>
-  >(
-    ...args: [K1, K2, K3, K4, K5, K6, K7, K8, ...(Part<any> | StoreSetter<any>)[]]
-  ): void;
-}
+export type WrappableNext<T, K extends Part<T>> = Exclude<Next<T, K>, NotWrappable>;
+
+type DistributeRest<T, P extends Part<T>> = P extends P ? [P, ...Rest<Next<T, P>>] : never;
```
Comment (bgoscinski) [inline packages/solid/store/src/store.ts:249]: Okay this one is a mystery for me. Can you point me where should I look for clues what `P extends P` does?

Comment (otonashixav) [reply packages/solid/store/src/store.ts:249]: It forces the type to be distributed. For example if `P = A | B`, the resultant type would be `[A, ...Rest<Next<T, A>>] | [B, ...Rest<Next<T, B>>]` instead of `[A | B, ...Rest<Next<T, A | B>>]`.

File: packages/solid/store/src/store.ts (hunk 1 above)
Comment (otonashixav) [reply packages/solid/store/src/store.ts]: I've removed this particular one. Since we're never going to pass a non-wrappable `T` to `Part`, there's no need to check and it improves performance which was previously a major concern.

**Comment by non25** [comment]
Hi, trying this out: https://playground.solidjs.com/?hash=2095550606&version=1.2.5
Can't think of a way to fix it :thinking: 
Works better in the playground I guess...

**Comment by atk** [comment]
The playground disables strict type checks, which breaks a few types that rely on them, unfortunately. Maybe we can enable strict in the playground, but selectively disable no-implicit-any and a few other convenient options that don't break strict types.

**Comment by otonashixav** [comment]
I think it isn't possible to avoid a type error in this case; the existing types also error here:
```ts
// interactions with generics
<T extends string>(v: T) => {
  type A = { a?: T; b?: Record<string, string>; c?: Record<T, string> };
  const [, setStore] = createStore<A>({});
  // should allow
  setStore("a", v);
  setStore("b", "a", "c");
  setStore("c", v, "c"); // <-- here
};

```

**Comment by atk** [comment]
Could `<T extends string = string>` work to narrow down the cases and still allow type overwrites?

**Comment by otonashixav** [comment]
It doesn't seem to work. Even if it did I'm not sure if it would be expected to the end user.

**Comment by atk** [comment]
The problem here is that T extends string can be an arbitrary union that doesn't fully overlap with keyof {} that is string | number | symbol. I don't think this is overly useful.

**Comment by otonashixav** [comment]
`T extends string` should be able to index `Record<T, string>` though, which is what's going on here. I suspect `Store<T>` is causing issues since I can index a non-store just fine:
```ts

  type A = { a?: T; b?: Record<string, string>; c?: Record<T, string> };
  const a = {} as A;
  const [store, setStore] = createStore<A>(a);
  // should allow
  setStore("a", v);
  setStore("b", "a", "c");
  // TODO
  setStore("c", v, "c"); // 
  store.c?.[v]; // <-- Expression produces a union type that is too complex to represent.ts(2590)
  a.c?.[v]; // <-- works
```

**Comment by atk** [comment]
Obviously, you cannot set a value that is readonly, so that error is to be expected. Maybe what we need is not DeepReadonly, but DeepStructuralReadonly, which only makes properties immutable that are arrays or objects.

**Comment by otonashixav** [comment]
It is not readonly; I'm not setting the value on the readonly store, but on a variable the same type as the value at the end of the store i.e. `let a: Readonly<Record<string, string>> = "1"` works, so this should too.

**Comment by otonashixav** [comment]
If no one is able to solve the type errors caused by [generics](https://github.com/solidjs/solid/pull/763/files#diff-718d3786c3e47b99ae8f43e84bf9d68788a8a193ac46f3e5e1d649d0963f6655R662), I believe no more improvements can be made until https://github.com/microsoft/TypeScript/issues/43122 is resolved.

**Comment by ryansolid** [comment]
I don't understand this stuff too well and there have been a few version.. but what is the state of this PR? Candidate for 1.3? What does it do?

**Comment by otonashixav** [comment]
I don't believet there's anything more to add to the PR, other than possibly a type for the return value of `createStore`. The main post includes most of the changes; the most noticable changes are likely `Store`, which now is fully readonly and simplified to allow ts to show better error messages, and `SetStoreFunction`, which has its types inferred much faster. 

I realise the one change I've left out is the replacement of instances of `property` typed as `string | number` with `keyof any`, resulting in a runtime change [here](https://github.com/solidjs/solid/pull/763/files#diff-86ed9ba88e2ceb505ab412a1b54f475b36555ad70a84af10685b8df8fa52f001R129), to avoid a runtime error in dev mode when wrapping a value whose property is a symbol. 

```ts
  const s = Symbol();
  const [store] = createStore({ [s]: { a: 1 } });
  const c = store[s].a; // <-- TypeError: Cannot convert a Symbol value to a string
```

**Comment by ryansolid** [comment]
Hmm.. I noticed type tests are off in the store package. I probably did that because I was having headaches with it. But a bunch of stuff is failing it might be worth using that to double-check the work here. This list should have "./store/test" added to it:
https://github.com/solidjs/solid/blob/main/packages/solid/tsconfig.test.json#L11

**Comment by otonashixav** [comment]
Some of the errors are expected, since there are some edge cases which cannot be represented with typescript currently. I suppose the tests shouldn't be failing though, so I'll change that. The other errors were missed due to `strict: false`, so those need fixing.

**Comment by yonathan06** [comment]
Could it be that this changes broke types for `setStore` using `produce`?
Getting this type check error:
Screen Shot 2022-02-01 at 10 40 46

[playground example](https://playground.solidjs.com/?hash=-187295914&version=1.3.3)

**Comment by atk** [comment]
I think the type for produce needs to unwrap the store.

**Comment by otonashixav** [comment]
I figured it was fine since there wasn't any issue with the test cases but the types are definitely wrong for the top level.

**Comment by otonashixav** [comment]
I see the issue: readonly on arrays isn't being removed by inference. See https://tsplay.dev/mL42vm for an example. Fix is in #824
